Move `PalletVersion` away from the crate version

bkchr opened the pull request:
Before this pr, `PalletVersion` was referring to the crate version that
hosted the pallet. This pr introduces a custom `package.metadata.frame`
section in the `Cargo.toml` that can contain a `pallet-version` key
value pair. While the value is expected to be a valid u16. If this
key/value pair isn't given, the version is set to 1.

It also changes the `PalletVersion` declaration. We now only have one
`u16` that represents the version. Not a major/minor/patch version. As
the old `PalletVersion` was starting with the `u16` major, decoding the
old values will work.

polkadot companion: https://github.com/paritytech/polkadot/pull/3526

shawntabrizi commented on frame/contracts/Cargo.toml:
```diff
@@ -77,3 +77,6 @@ try-runtime = ["frame-support/try-runtime"]
 unstable-interface = []
+
+[package.metadata.frame]
+pallet-version = 4
```
If we store this item in the pallet as `stoarge-version`, i think the name matching here would make sense.

```suggestion
storage-version = 4
```

andresilva replied:
Should this be `3` to match the existing pallet version?

bkchr replied:
I can do this, but then I would rename the struct etc to `StorageVersion` as well?

thiolliere replied:
we use pallet version everywhere, in for `PalletVersion::storage_version` should be understood the pallet version in storage.

thiolliere replied:
> Should this be `3` to match the existing pallet version?

I'm not sure, in the code if version is 3 we do an upgrade, I guess next release of contract should bump the version in order not to trigger the upgrade on every runtime upgrade. (though this runtime upgrade is harmless)

bkchr replied:
> we use pallet version everywhere, for me `PalletVersion::storage_version` should be understood the pallet version in storage.

So what you is your favorite naming?

athei replied:
I think the naming should make clear that it is about **storage** compatibility. `PalletVersion` simply makes me ask myself why it has to exist when we have the perfectly fine crate version. If you ask me I would completely remove everything that refers to it as a pallet version:

`StorageVersion::current_version(), StorageVersion::stored_version()`.

thiolliere replied:
> So what you is your favorite naming?

considering we always refer to it as pallet version, and that I understand `PalletVersion::storage_version` as the pallet version in storage, then my favorite is to use `pallet-version` in the Cargo.toml here.

But if we rename everything to storage-version I don't mind either.

athei commented:
Why is this useful and to which version (crate or pallet) do I refer to in my `CHANGELOG.md` from now on?

kianenigma commented on frame/support/procedural/src/pallet_version.rs:
```diff
@@ -19,46 +19,59 @@
+				)
+			}
+		} else {
+			1
```
This value is supposed to be stored in storage. My default would be that this macro would similarly return `0` if none is defined.

apopiak replied:
Agreed that the default value should be 0.

thiolliere reviewed (dismissed):
code looks good to me,

I don't have strong opinion between storage-version and pallet-version.
* storage-version makes it more close to its current usage: we bump it when storage is breaking and do some conditional with it in storage migrations
* pallet-version makes it more close to its original goal (as I understand it): having a stored number which can identifier clearly the pallet. So we can easily checkout the pallet code. But if we don't bump this number when the pallet behaviour changes, then the name is more misleading than helpful.

Also I agree 0 could make more sense when no version is defined https://github.com/paritytech/substrate/pull/9165/files#r655954312

apopiak commented:
Just FYI that this XCM PR currently assumes (major, minor, patch) for pallet versions:
https://github.com/paritytech/xcm-format/pull/22

bkchr commented:
> Just FYI that this XCM PR currently assumes (major, minor, patch) for pallet versions:
> [paritytech/xcm-format#22](https://github.com/paritytech/xcm-format/pull/22)

Hmm good point. Yeah, I don't know. I'm still seeing the "need" for this, but currently no one bumps any versions anyway... So, yeah... It is hard.

bkchr commented:
Or we just keep the `PalletVersion` with `major`, `minor` and `patch` and make these values set able through the `Cargo.toml` as I have done it here already in the pr?

bkchr commented:
@kianenigma @thiolliere @athei what do you think?

kianenigma commented:
TBH I am only interested in this for storage migrations (see my linked issue), so that we can get rid of all the transient storage items. For that, a single version is enough. If the triplet is also useful in certain cases, we can

```
version = "x.y.z" // This is written as 'PalletVersion'

[package.metadata.frame]  
storage-version = "v" // This is written as `StorageVersion`
```

`version` is only about crates.io and semver. `storage-version` is only about the version of storage, and with each change, it should be bumped. In an ideal world, any `storage-version` bump must have a major `version` update as well. But the opposite is not true.

shawntabrizi commented:
I think this should be named `storage-version` and contain just a single value, similar to `spec_version` and stuff

athei commented:
I agree with shawn and kian. We only need one number because this version seems to be only about storage (for everything else we have the crate version). We should name is `storage-version` and it should be a single number.

apopiak commented:
Agree as well. Just linked the XCM PR to make sure that we adjust that once we change this.

bkchr commented:
Okay I have rewritten the entire thing.

- PalletVersion was removed and replaced by StorageVersion
- We don't have any magical macro anymore that pulls data out of the Cargo.toml. I decided for making the storage version explicit as a const in the top of the crate. However, we still need to tell the pallet macros about it.
- The pallet author is also now required to set the correct storage version after a migration.
- I added a migration for migrating from PalletVersion to StorageVersion.

shawntabrizi commented on frame/support/test/tests/pallet_instance.rs:
```diff
@@ -567,19 +557,9 @@ fn pallet_hooks_expand() {
 #[test]
 fn pallet_on_genesis() {
```
What does this test now that you removed all the asserts?

bkchr replied:
First, we always now have a version returned and if there is nothing in the storage we return the default value (0).

Secondly the almost same test exists in `pallet.rs`

shawntabrizi replied:
Am I going crazy? Or do I see the diff wrong?

All I see is this:

```
pallet::Pallet::<Runtime>::on_genesis();
pallet::Pallet::<Runtime, pallet::Instance1>::on_genesis();
```

This only shows me these two calls dont panic?

bkchr replied:
Yeah, maybe I should remove this xD

shawntabrizi commented on frame/support/test/tests/pallet.rs:
```diff
@@ -839,15 +837,54 @@ fn pallet_hooks_expand() {
 		assert_eq!(
-			pallet::Pallet::<Runtime>::storage_version(),
-			Some(pallet::Pallet::<Runtime>::current_version()),
+			pallet::Pallet::<Runtime>::current_storage_version(),
+			pallet::Pallet::<Runtime>::active_storage_version(),
```
A personal nit pick, I think these two names can be too easily confused with one another.

I would want to name one `on_chain_storage_version`

bkchr replied:
I'm fine :D  I'm bad at naming :P 

So, active to `on_chain_storage_version`?

shawntabrizi replied:
yup!

shawntabrizi reviewed (approved):
this makes sense to me

shawntabrizi commented on frame/support/procedural/src/pallet/expand/hooks.rs:
```diff
@@ -40,19 +40,17 @@ pub fn expand_hooks(def: &mut Def) -> proc_macro2::TokenStream {
 		quote::quote! {
 			#frame_support::log::info!(
 				target: #frame_support::LOG_TARGET,
-				"⚠️ {} declares internal migrations (which *might* execute), setting storage version to {:?}",
```
is it not possible to still output the current and new storage version?

bkchr replied:
As we don't update the version automatically in the storage, I'm not sure what the benefit of the message is.

kianenigma replied:
If possible, it is still benefical to display it.

kianenigma replied:
it be a violation of `current_storage_version` and `on_chain_storage_version` are not the same in most cases, I think both should be displayed here.

kianenigma replied:
@bkchr ^^

thiolliere replied:
seems fixed

thiolliere replied:
seems fixed

shawntabrizi commented on frame/support/src/dispatch.rs:
```diff
@@ -1446,25 +1444,14 @@ macro_rules! decl_module {
 
 				$crate::log::info!(
 					target: $crate::LOG_TARGET,
-					"⚠️ {} declares internal migrations (which *might* execute), setting storage version to {:?}",
```
same note here about storage version being displayed is nice to have

thiolliere replied:
seems fixed

shawntabrizi commented on frame/contracts/src/migration.rs:
```diff
@@ -16,24 +16,20 @@
+			b"",
+		);
+
+		StorageVersion::new(4).put::<Pallet<T>>();
```
Why do we need this line? Wont it be overwritten by the other storage version `put` in `migrate()`?

bkchr replied:
As written in my message above, we don't write the storage version automatically anymore into the storage.

kianenigma replied:
While it feels like a step backwards, I agree that being explicit is better in this case, and it least for now.

kianenigma replied:
ambivalent, feels like a step backwards. Having it automatically written per `runtime_upgrade` is always what you want to do.

Basically you are allowing `pallet_storage_version` and `current_storage_version` to go out of sync. When is this useful?

bkchr replied:
> Basically you are allowing `pallet_storage_version` and `current_storage_version` to go out of sync. When is this useful?

This is not useful. However, if you don't have executed the migration your fucked anyway.

IDK. I can bring the automatic update back. This way it is less magic. Especially useful for when we finally ban pallet `on_runtime_upgrade`, because we don't need to be afraid that the version is updated before the migration is executed in the runtime file.

kianenigma replied:
okay lets keep it manual for now then, less "magic".

athei commented on frame/contracts/src/lib.rs:
```diff
@@ -134,6 +134,9 @@ type BalanceOf<T> =
 	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::NegativeImbalance;
 
+/// The current storage version.
+pub const STORAGE_VERSION: StorageVersion = StorageVersion::new(4);
```
Why do we need this public const? Users can determine the version by using `trait GetStorageVersion`, right? This introduces a possibility for a mismatch: The pallet programmer uses a different value in the annotation.

athei commented on frame/support/procedural/Cargo.toml:
```diff
@@ -20,6 +20,7 @@ proc-macro2 = "1.0.6"
 quote = "1.0.3"
 Inflector = "0.11.4"
 syn = { version = "1.0.58", features = ["full"] }
+cargo_metadata = "0.13.1"
```
We no longer need this, right?

kianenigma commented on frame/support/src/migrations.rs:
```diff
@@ -0,0 +1,60 @@
+		sp_io::storage::clear(&pallet_version_key(<T as PalletInfoAccess>::name()));
+
+		let version = <T as GetStorageVersion>::current_storage_version();
+		version.put::<T>();
```
would be nice if you log what pallet version is being replaced with what storage version

bkchr replied:
Would require again rather useless code. You can just check which storage version will be set, by checking the pallets.

kianenigma reviewed (commented):
has a few open review comments, like https://github.com/paritytech/substrate/pull/9165/files#r674533695. But should be fine after they're fixed.

kianenigma reviewed (approved):
LGTM

thiolliere commented on frame/support/src/dispatch.rs:
```diff
@@ -2023,33 +2168,11 @@ macro_rules! decl_module {
-				$crate::crate_to_pallet_version!()
-					.put_into_storage::<<$trait_instance as $system::Config>::PalletInfo, Self>();
-			}
+			fn on_genesis() {}
```
I think we should do same as in pallet macro: on genesis the storage version is known: it is the storage version of the pallet. So it should be safe to write it.

```suggestion
			fn on_genesis() {
				let storage_version = <Self as $crate::traits::GetStorageVersion>::current_storage_version();
				storage_version.put::<Self>();
			}
```

bkchr replied:
Good find!

thiolliere commented on frame/support/src/traits/metadata.rs:
```diff
@@ -125,47 +115,75 @@ impl PalletVersion {
-	/// If there was no previous version of the pallet stored in the state,
-	/// this function returns `None`.
-	fn storage_version() -> Option<PalletVersion>;
+/// It differentiates between current and active storage version. Both should be only out of sync
+/// when a new runtime upgrade was applied and the runtime migrations did not yet executed.
+/// Otherwise it means that the pallet works with an unsupported storage version and unforeseen
+/// stuff can happen.
+///
+/// The current storage version is the version of the pallet as supported at runtime. The active
+/// storage version is the version of the pallet in the storage.
+///
+/// It is required to update the active storage version manually when a migration was applied.
```
using the same name as functions makes it more easy to grasp maybe
```suggestion
/// It differentiates between current and on-chain storage version. Both should be only out of sync
/// when a new runtime upgrade was applied and the runtime migrations did not yet executed.
/// Otherwise it means that the pallet works with an unsupported storage version and unforeseen
/// stuff can happen.
///
/// The current storage version is the version of the pallet as supported at runtime. The active
/// storage version is the version of the pallet in the storage.
///
/// It is required to update the on-chain storage version manually when a migration was applied.
```

thiolliere replied:
we could also link to `StorageVersion::put` for manually inserting the version.

thiolliere reviewed (approved):
needs to fix the on genesis for decl_storage macro: https://github.com/paritytech/substrate/pull/9165#discussion_r677594576

But otherwise it is good to me

thiolliere commented:
Note for other about migration:

The storage version is now stored under a different prefix than the old pallet version. To upgrade you can use `migrate_from_pallet_version_to_storage_version` like in https://github.com/paritytech/polkadot/pull/3526/files
Basically adding the struct:
```rust
/// Migrate from `PalletVersion` to the new `StorageVersion`
pub struct MigratePalletVersionToStorageVersion;

impl OnRuntimeUpgrade for MigratePalletVersionToStorageVersion {
	fn on_runtime_upgrade() -> frame_support::weights::Weight {
		frame_support::migrations::migrate_from_pallet_version_to_storage_version::<AllPalletsWithSystem>(
				&RocksDbWeight::get()
		)
	}
}
```
And add this struct to the generic `CustomOnRuntimeUpgrade` of `Executive`

Warning: The custom runtime upgrade happens before the pallet's runtime upgrede (defined inside hooks). Thus any hook who rely on the pallet version won't be triggered and migration must be done manually. in substrate repo, pallet-contracts is the only pallet which look into the stored pallet version.

bkchr commented:
@thiolliere ty

bkchr commented:
bot merge
